README.md
Better Transbank SDK
====================

Un SDK no oficial (¡pero mejor!) de los servicios de pago de Transbank.

[![Actions Status](https://github.com/better-transbank/sdk/workflows/CI/badge.svg)](https://github.com/better-transbank/sdk/actions)

## Instalación
Puedes instalar este SDK de manera sencilla con composer

```
composer require better-transbank/sdk
```

> **PROMESA DE ESTABILIDAD**:
> 
> Esta librería adhiere a [versionamiento semántico](https://semver.org/lang/es/) por lo
> que prometemos mantener la estabilidad de la api en versiones minor y patch desde
> `1.0.0` en adelante.

## Inicio Rápido
Crear una transacción normal y mostrar el formulario de pago de WebpayPlus es extremadamente sencillo:

```php
use BetterTransbank\SDK\Config;
use BetterTransbank\SDK\Html\PaymentForm;
use BetterTransbank\SDK\Services\WebpayPlus\Transaction;
use BetterTransbank\SDK\TestingCredentials;
use BetterTransbank\SDK\Transbank;

$config = Config::fromCredentials(TestingCredentials::forWebpayPlusNormal());
$transbank = Transbank::create($config);

$transaction = Transaction::normal(
    '12345',
    10000,
    'http://localhost:8000/return',
    'http://localhost:8000/final'
);

$result = $transbank->webpayPlus()->register($transaction);
PaymentForm::prepare($result)->send();
```

> **NOTA:** Evita usar el método `PaymentForm::send()` ya que presupone que usas PHP con CGI y
> que tienes múltiples scripts php en tu codebase.
> En vez de eso, recomendamos usar buenas abstracciones HTTP como `symfony/http-foundation`
> o cualquier implementación de `psr/http-message` y contar con un solo entrypoint.

## ¿Por qué un nuevo SDK?
Por mucho tiempo he usado el SDK oficial de Webpay y he encontrado diversos problemas con él.

- El protocolo SOAP se derrama sobre la insipiente abstracción de los diferentes servicios
- La firma WSSE del XML se realiza con un set de clases arcanas que ya no tienen soporte (y que no tienen licencia MIT)
- Carece de buenas prácticas de programación orientada a objetos (encapsulamiento, coding to an interface, patrones de diseño)
- Carece de tipado estricto en casi todos los lugares
- Tiene una api verbose y compleja
- Carece de suficientes pruebas unitarias
- No documenta las posibles excepciones en el DocBlock
- No utiliza las nuevas funcionalidades y estándares presentes en PHP 7+

Por estas razones decidí desarrollar un nuevo SDK que se ajustara a mis necesidades y que fuera muchísimo
más confiable y fácil de usar.

Para saber más en detalle las razones que me llevaron a desarrollar este SDK, puedes ver [la entrada de mi blog]
donde también explico todo el proceso de desarrollo en detalle.

> **NOTA:** Freshwork Studio desarrolló un SDK que soluciona algunos de estos problemas, pero en esencia es un wrapper
con un poco de funcionalidad extra sobre la misma lógica del SDK oficial. Esta, en cambio, es una implementación
desde cero.

## Servicios de Transbank Implementados
- [x] Webpay Plus (Transacciones Normales, Mútiples y de Subscripción)
- [x] Commerce Integration (Captura Diferida y Anulación)
- [ ] Webpay OneClick

## Documentación
Esta libreria dispone de una [documentación detallada] que puedes consultar para conocer la librería
en profundidad.

## Contribuye
Puedes leer la guía para contribuir a este repositorio, en donde se explica cómo contribuir con código o reportando issues.

Además, si esta librería o alguna otra de las que he desarrollado te ha servido y sacado de un apuro,
puedes comprame una cervecita o un café por Github Sponsors.

[la entrada de mi blog]: https://mnavarro.dev/2020/a-brand-new-transbank-sdk/
[documentación detallada]: https://better-transbank.mnavarro.dev

CHANGELOG.md
# Changelog
Todos los cambios notables en este proyecto están documentados en este archivo.

El formato está basado en [Keep a Changelog](https://keepachangelog.com/es-ES/1.0.0/),
y este proyecto adhiere a [Versionamiento Semántico](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.4.0] - 2020-06-07

## Modified
- Modificada la clase transacción y la api del servicio de Webpay Plus.
- Se cambiaron namespaces de casi todas las clases y servicios
- Se quitaron los repositorios de certificados y se centralizaron en `better-transbank/certificates`.
- Se quitó el CI de la rama master, debido a que es redundante
- Se quitó la documentación porque será colocada en su propio repositorio `better-transbank/docs`
- Quitadas las comparaciones con estilo Yoda

## Added
- Agregado comando de consola para generar certificados usando `oo-php/openssl`
- Agregado soporte para servicio Webpay Commerce
- Integración Continua de cambios en código, con pruebas unitarias y manejo de estilo
- Psalm fue añadido al pipeline y ya no muestra errores de tipado en el código
- Dependencia a `better-transbank/certificates` como repositorio de certificados

## [0.3.0] - 2020-01-09 

### Changed
- Movida la lógica de formado WSSE fuera de `TransbankSoapClient` y encapsulada en la nueva clase `WSSE\RequestDocument`.

### Added
- Creada la clase `WSSE\ResponseDocument` que encapsula la validación de una respuesta XML de Transbank.

### Fixed
- Corregido el objeto `Details` dentro de `TransactionResult`. Ahora, puede contener un arreglo de details.
Esto soporta Webpay Mall con varias transacciones.

## [0.2.0] - 2020-01-08

### Added
- Webpay Plus Pat Pass
- Ejemplos extra del SDK para Webpay Normal en todos sus modos (Normal, Mall y PatPass)

### Changed
- Las clases `Certificate` y `PrivateKey` ahora abren los archivos al momento de instanciación, para
ahorrar I/O en las operaciones.

### Deprecated
- El signature del constructor de la clase `Credentials` ha sido deprecado. A partir de 1.0.0 ya
 no soportará nombres de archivo como parámetros, sino las instancias `Certificate` y `PrivateKey`. Como
 reemplazo para instanciación a través de nombres de archivo, se disponibiliza el método estático
 `Credentials::fromFilesPath()`.

## [0.1.0] - 2020-01-08
 
### Added
- Añadido el módulo `Soap` que contiene el `TransbankSoapClient` que firma XML usando los `Credentials` provistos. 
- Definido `WebpayClient`
- Implementado `SoapWebpayClient`
- Implementado `Psr14WebpayClient`
- Añadidos los helpers para respuestas fáciles en html
- Añadido CS-Fixer, Psalm y PHPUnit
- Añadido el cliente SOAP con Logger
- Añadido Webpay decorado con Event Dispatcher
 
[Unreleased]: https://github.com/better-transbank/sdk/compare/0.4.0...HEAD
[0.4.0]: https://github.com/better-transbank/sdk/compare/0.3.0...0.4.0
[0.3.0]: https://github.com/better-transbank/sdk/compare/0.2.0...0.3.0
[0.2.0]: https://github.com/better-transbank/sdk/compare/0.1.0...0.2.0
[0.1.0]: https://github.com/better-transbank/sdk/compare/releases/tag/0.1.0

examples/wp-patpass.php
<?php
declare(strict_types=1);

use BetterTransbank\SDK\Config;
use BetterTransbank\SDK\Html\PaymentForm;
use BetterTransbank\SDK\Html\RedirectView;
use BetterTransbank\SDK\Services\WebpayPlus\Customer;
use BetterTransbank\SDK\Services\WebpayPlus\Subscription;
use BetterTransbank\SDK\Services\WebpayPlus\Transaction;
use BetterTransbank\SDK\TestingCredentials;
use BetterTransbank\SDK\Transbank;

require_once __DIR__ . '/../vendor/autoload.php';

// This example shows how you can use the sdk in a CGI single-script context.
// You can run this example with PHP built-in web server:
// php -S 0.0.0.0:8000 -t examples/ examples/wp-patpass.php

$config = Config::fromCredentials(TestingCredentials::forWebpayPlusSubscription());
$transbank = Transbank::create($config);

if ($_SERVER['REQUEST_METHOD'] === 'GET') {


    $customer = Customer::register(
        '11.111.111-2',
        'Juanito',
        'Pérez',
        'González',
        'jpergon@example.com',
        '123456789'
    );

    $subscription = Subscription::define(
        '12345',
        10000,
        '124214125',
        new DateTime('+1 year')
    );

    $transaction = Transaction::subscription(
        $subscription,
        $customer,
        'commerce@example.com',
        'http://localhost:8000/return',
        'http://localhost:8000/final'
    );

    $result = $transbank->webpayPlus()->register($transaction);
    PaymentForm::prepare($result)->send();
    exit;
}

if ($_SERVER['REQUEST_URI'] === '/return') {
    $token = $_REQUEST['token_ws'];
    $transactionInfo = $transbank->webpayPlus()->info($token);
    if ($result->isSuccessful()) {
        $transbank->webpayPlus()->confirm($token);
    }
    RedirectView::prepare($result, $token)->send();
    exit;
}

if ($_SERVER['REQUEST_URI'] === '/final') {
    echo 'Transaction OK';
    exit;
}